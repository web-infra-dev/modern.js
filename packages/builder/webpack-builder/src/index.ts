export { createBuilder } from './core/createBuilder';
export { mergeBuilderConfig } from './shared/utils';

export type { BuilderInstance } from './core/createBuilder';
export type {
  // Plugin Types
  BuilderPlugin,
  BuilderContext,
  BuilderOptions,
  BuilderPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  OutputConfig,
  SourceConfig,
  BuilderConfig,
  SecurityConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSLoaderOptions,
  CssExtractOptions,
  LessLoaderOptions,
  SassLoaderOptions,
  HTMLPluginOptions,
  StyleLoaderOptions,
  AutoprefixerOptions,
  TerserPluginOptions,
  PostCSSLoaderOptions,
} from './types';
