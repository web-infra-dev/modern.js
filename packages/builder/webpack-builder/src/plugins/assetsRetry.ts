import { BuilderPlugin } from '../types/plugin';

export function PluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'webpack-builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getBuilderConfig();
        if (!config.output?.assetsRetry || target === 'node') {
          return;
        }
        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
