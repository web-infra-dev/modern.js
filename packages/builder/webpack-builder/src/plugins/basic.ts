import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of webpack
 */
export const PluginBasic = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-basic',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      chain.context(api.context.rootPath);

      chain.mode(isProd ? 'production' : 'development');

      chain.merge({
        infrastructureLogging: {
          // Using `error` level to avoid `webpack.cache.PackFileCacheStrategy` logs
          level: 'error',
        },
      });

      // if the chunk size exceeds 1MiB, we will throw a warning
      chain.performance.maxAssetSize(1024 * 1024);
      chain.performance.maxEntrypointSize(1024 * 1024);

      // This will be futureDefaults in webpack 6
      chain.module.parser.set('javascript', {
        exportsPresence: 'error',
      });
    });
  },
});
