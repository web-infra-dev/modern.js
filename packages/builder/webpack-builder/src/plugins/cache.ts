import { join } from 'path';
import { isFileExists } from '../shared';
import type { BuilderPlugin } from '../types';

export const PluginCache = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-cache',

  setup(api) {
    api.modifyWebpackChain(async (chain, { target, env }) => {
      const { context } = api;
      const cacheDirectory = join(context.cachePath, 'webpack');
      const rootPackageJson = join(context.rootPath, 'package.json');
      const browserslistConfig = join(context.rootPath, '.browserslistrc');

      /**
       * webpack can't detect the changes of framework config, tsconfig and browserslist config.
       * but they will affect the compilation result, so they need to be added to buildDependencies.
       */
      const buildDependencies: Record<string, string[]> = {
        packageJson: [rootPackageJson],
      };
      if (context.configPath) {
        buildDependencies.config = [context.configPath];
      }
      if (context.tsconfigPath) {
        buildDependencies.tsconfig = [context.tsconfigPath];
      }
      if (await isFileExists(browserslistConfig)) {
        buildDependencies.browserslistrc = [browserslistConfig];
      }

      chain.cache({
        // The default cache name of webpack is '${name}-${env}', and the `name` is `default` by default.
        // We set cache name to avoid cache conflicts of different targets.
        name: `${target}-${env}`,
        type: 'filesystem',
        cacheDirectory,
        buildDependencies,
      });
    });
  },
});
