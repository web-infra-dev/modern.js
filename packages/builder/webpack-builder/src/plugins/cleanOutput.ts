import type { BuilderPlugin } from '../types';

export const PluginCleanOutput = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-clean-output',

  setup(api) {
    const config = api.getBuilderConfig();

    if (config.output?.cleanDistPath) {
      const clean = async () => {
        const { emptyDir } = await import('@modern-js/utils');
        const { distPath } = api.context;
        await emptyDir(distPath);
      };

      api.onBeforeBuild(clean);
      api.onBeforeStartDevServer(clean);
    }
  },
});
