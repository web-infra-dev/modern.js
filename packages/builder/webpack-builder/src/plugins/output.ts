import { getDistPath, getFilename, DEFAULT_PORT } from '../shared';
import type {
  BuilderConfig,
  BuilderContext,
  BuilderPlugin,
  MiniCSSExtractPluginOptions,
} from '../types';

function getPublicPath({
  config,
  isProd,
  context,
}: {
  config: BuilderConfig;
  isProd: boolean;
  context: BuilderContext;
}) {
  const { dev, output } = config;

  let publicPath = '/';

  if (isProd) {
    if (output?.assetPrefix) {
      publicPath = output.assetPrefix;
    }
  } else if (typeof dev?.assetPrefix === 'string') {
    publicPath = dev.assetPrefix;
  } else if (dev?.assetPrefix === true) {
    const ip = context.devServer?.ip || 'localhost';
    const port = context.devServer?.port || DEFAULT_PORT;
    publicPath = `//${ip}:${port}/`;
  }

  if (!publicPath.endsWith('/')) {
    publicPath += '/';
  }

  return publicPath;
}

export const PluginOutput = (): BuilderPlugin => ({
  name: 'webpack-builder-plugin-output',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd, isServer, CHAIN_ID }) => {
      const config = api.getBuilderConfig();
      const jsPath = getDistPath(config, 'js');
      const cssPath = getDistPath(config, 'css');

      const publicPath = getPublicPath({
        config,
        isProd,
        context: api.context,
      });
      const enableExtractCSS = !config.tools?.styleLoader;

      // js output
      const jsFilename = getFilename(config, 'js', isProd);
      chain.output
        .path(api.context.distPath)
        .filename(`${jsPath}/${jsFilename}`)
        .chunkFilename(`${jsPath}/async/${jsFilename}`)
        .publicPath(publicPath)
        // disable pathinfo to improve compile performance
        // the path info is useless in most cases
        // see: https://webpack.js.org/guides/build-performance/#output-without-path-info
        .pathinfo(false)
        // since webpack v5.54.0+, hashFunction supports xxhash64 as a faster algorithm
        // which will be used as default when experiments.futureDefaults is enabled.
        .hashFunction('xxhash64');

      // css output
      if (enableExtractCSS) {
        const { default: MiniCssExtractPlugin } = await import(
          'mini-css-extract-plugin'
        );
        const { applyOptionsChain } = await import('@modern-js/utils');
        const extractPluginOptions = applyOptionsChain<
          MiniCSSExtractPluginOptions,
          null
        >({}, config.tools?.cssExtract?.pluginOptions || {});

        const cssFilename = getFilename(config, 'css', isProd);

        chain
          .plugin(CHAIN_ID.PLUGIN.MINI_CSS_EXTRACT)
          .use(MiniCssExtractPlugin, [
            {
              filename: `${cssPath}/${cssFilename}`,
              chunkFilename: `${cssPath}/async/${cssFilename}`,
              ignoreOrder: true,
              ...extractPluginOptions,
            },
          ]);
      }

      // ssr output
      if (isServer) {
        const { SERVER_BUNDLE_DIRECTORY } = await import(
          '@modern-js/utils/constants'
        );
        const filename = `${SERVER_BUNDLE_DIRECTORY}/[name].js`;

        chain.output
          .filename(filename)
          .chunkFilename(filename)
          .libraryTarget('commonjs2');
      }
    });
  },
});
