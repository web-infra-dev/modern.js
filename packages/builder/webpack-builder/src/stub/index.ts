export { createStubBuilder } from './builder';
export { createStubContext } from './context';
