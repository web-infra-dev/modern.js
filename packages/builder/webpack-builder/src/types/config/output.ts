import type { CopyPluginOptions } from '../thirdParty';

export type DistPathConfig = {
  root?: string;
  js?: string;
  css?: string;
  svg?: string;
  font?: string;
  html?: string;
  image?: string;
  media?: string;
};

export type FilenameConfig = {
  js?: string;
  css?: string;
  svg?: string;
  font?: string;
  image?: string;
  media?: string;
};

export type DataUriLimit = {
  svg?: number;
  font?: number;
  image?: number;
  media?: number;
};

export type Polyfill = 'usage' | 'entry' | 'ua' | 'off';

export type AssetsRetryHookContext = {
  times: number;
  domain: string;
  url: string;
  tagName: string;
};

export type AssetsRetryOptions = {
  type?: string[];
  domain?: string[];
  max?: number;
  test?: string | ((url: string) => boolean);
  crossOrigin?: boolean;
  onRetry?: (options: AssetsRetryHookContext) => void;
  onSuccess?: (options: AssetsRetryHookContext) => void;
  onFail?: (options: AssetsRetryHookContext) => void;
};

export interface OutputConfig {
  copy?: CopyPluginOptions | CopyPluginOptions['patterns'];
  distPath?: DistPathConfig;
  filename?: FilenameConfig;
  polyfill?: Polyfill;
  assetPrefix?: string;
  dataUriLimit?: number | DataUriLimit;
  cleanDistPath?: boolean;
  disableMinimize?: boolean;
  disableSourceMap?: boolean;
  disableFilenameHash?: boolean;
  disableInlineRuntimeChunk?: boolean;
  enableAssetManifest?: boolean;
  enableAssetFallback?: boolean;
  enableLatestDecorators?: boolean;
  enableCssModuleTSDeclaration?: boolean;
  enableInlineScripts?: boolean;
  enableInlineStyles?: boolean;
  overrideBrowserslist?: string[];
  svgDefaultExport?: 'component' | 'url';
  assetsRetry?: AssetsRetryOptions;
}
