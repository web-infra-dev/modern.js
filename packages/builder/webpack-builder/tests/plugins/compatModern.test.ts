import { expect, describe, it } from 'vitest';
import { PluginCompatModern } from '../../src/plugins/compatModern';
import { PluginFallback } from '../../src/plugins/fallback';
import { PluginManifest } from '../../src/plugins/manifest';
import { PluginModuleScopes } from '../../src/plugins/moduleScopes';
import { PluginMoment } from '../../src/plugins/moment';
import { PluginOutput } from '../../src/plugins/output';
import { PluginResolve } from '../../src/plugins/resolve';
import { createStubBuilder } from '../../src/stub';

describe('plugins/compatModern', () => {
  // skipped because this case time out in CI env
  it.skip('should apply compatible webpack configs correctly', async () => {
    const builder = await createStubBuilder({
      plugins: [
        PluginOutput(),
        PluginResolve(),
        PluginMoment(),
        PluginManifest(),
        PluginCompatModern(),
        PluginModuleScopes(),
        PluginFallback(),
      ],
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should apply name and extensions for node target correctly', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCompatModern()],
      target: ['node'],
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config.name).toEqual('server');
    expect(config.resolve?.extensions).toEqual([
      '.node.tsx',
      '.node.ts',
      '.node.jsx',
      '.node.js',
    ]);
  });

  it('should apply name for modern target correctly', async () => {
    const builder = await createStubBuilder({
      plugins: [PluginCompatModern()],
      target: ['modern-web'],
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config.name).toEqual('modern');
  });
});
