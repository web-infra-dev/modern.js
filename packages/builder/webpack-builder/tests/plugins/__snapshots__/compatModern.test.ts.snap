// Vitest Snapshot v1

exports[`plugins/compatModern > should apply compatible webpack configs correctly 1`] = `
{
  "module": {
    "rules": [
      {
        "resolve": {
          "fullySpecified": false,
        },
        "test": /\\\\\\.m\\?js/,
      },
      {
        "oneOf": [
          {
            "exclude": [
              /\\\\\\.\\(js\\|mjs\\|cjs\\|jsx\\)\\$/,
              /\\\\\\.\\(ts\\|mts\\|cts\\|tsx\\)\\$/,
              /\\\\\\.html\\$/,
              /\\\\\\.json\\$/,
            ],
            "type": "asset/resource",
          },
        ],
      },
    ],
  },
  "name": "client",
  "output": {
    "chunkFilename": "static/js/async/[name].js",
    "filename": "static/js/[name].js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/packages/builder/webpack-builder/dist",
    "publicPath": "/",
  },
  "plugins": [
    MiniCssExtractPlugin {
      "_sortedModulesCache": WeakMap {},
      "options": {
        "chunkFilename": "static/css/[name].css",
        "experimentalUseImportModule": undefined,
        "filename": "static/css/[name].css",
        "ignoreOrder": true,
        "runtime": true,
      },
      "runtimeOptions": {
        "attributes": undefined,
        "insert": undefined,
        "linkType": "text/css",
      },
    },
    IgnorePlugin {
      "checkIgnore": [Function],
      "options": {
        "contextRegExp": /moment\\$/,
        "resourceRegExp": /\\^\\\\\\.\\\\/locale\\$/,
      },
    },
    WebpackManifestPlugin {
      "options": {
        "assetHookStage": Infinity,
        "basePath": "",
        "fileName": "asset-manifest.json",
        "filter": null,
        "generate": [Function],
        "map": null,
        "publicPath": "/",
        "removeKeyHash": /\\(\\[a-f0-9\\]\\{16,32\\}\\\\\\.\\?\\)/gi,
        "seed": undefined,
        "serialize": [Function],
        "sort": null,
        "transformExtensions": /\\^\\(gz\\|map\\)\\$/i,
        "useEntryKeys": false,
        "useLegacyEmit": false,
        "writeToFileEmit": false,
      },
    },
  ],
  "resolve": {
    "extensions": [
      ".web.mjs",
      ".mjs",
      ".web.js",
      ".js",
      ".web.jsx",
      ".jsx",
      ".web.json",
      ".json",
    ],
  },
}
`;
